allennlp/modules/seq2seq_encoders/efficient_multi_head_attention.py

```diff
@@ -0,0 +1,159 @@
+import torch
+
+from torch.autograd import Variable
+from torch.nn import Dropout, Linear
+from torch.nn import Parameter
+from torch.nn import init
+
+from allennlp.nn.util import last_dim_softmax
+from allennlp.modules.seq2seq_encoders.seq2seq_encoder import Seq2SeqEncoder
+
+
+@Seq2SeqEncoder.register("efficient_multi_head_attention")
+class EfficientMultiHeadAttention(Seq2SeqEncoder):
```

Comment:
I'd probably just call this `MultiHeadSelfAttention`.  We should make the other case the one with the longer name, and make it clear it should only be used when you want to explore similarity options (and we can do that change in a different PR).